classes/Models/Services/Tokens.php

```diff
@@ -0,0 +1,150 @@
+
+        // Check that expected token isn't expired.
+        $now = new \DateTime();
+        $expires = new \DateTime($expiresOn);
+        if ($expires < $now) {
+            throw new BadRequestHttpException('The API Token has expired.');
+        }
+
+        // finally check that the provided token matches it's stored hash.
+        if (!password_verify($password, $expectedToken)) {
+            throw new AccessDeniedHttpException('Invalid API token.');
+        }
+
+        // and if we've made it this far we can safely return the requested Users data.
+        return XDUser::getUserByID($dbUserId);
+    }
+
+    /**
+     * This function is a stop-gap that is meant to be used to protect controller endpoints until they can be moved to
+     * the new REST stack.
+     *
+     * @return XDUser|null if the authentication is successful then an XDUser instance for the authenticated user will
+     * be returned, if the authentication is not successful then null will be returned.
+     */
+    public static function authenticateToken()
+    {
+
+        $rawToken = self::getRawToken();
+        if (empty($rawToken)) {
+            // we want to the token authentication to be optional so instead of throwing an exception we return null.
+            // This allows us to provide token authentication to existing endpoints without impeding their normal use.
+            return null;
+        }
+
+        // We expect the token to be in the form /^(\d+).(.*)$/ so just make sure it at least has the required delimiter.
+        $delimPosition = strpos($rawToken, Tokens::DELIMITER);
+        if ($delimPosition === false) {
+            // Same as above, token authentication is optional so we return null instead of throwing an exception.
+            return null;
+        }
+
+        $userId = substr($rawToken, 0, $delimPosition);
+        $token = substr($rawToken, $delimPosition + 1);
+
+        try {
+            return Tokens::authenticate($userId, $token);
+        } catch (Exception $e) {
+            // and again, same as above.
+            return null;
+        }
+    }
+
+    /**
+     * Attempt to retrieve the raw API Token from one of the following sources:
+     *   - Headers
+     *   - GET Parameters
+     *   - POST Parameters
+     *
+     * @return null|string returns the api token if found else it returns null.
+     */
+    private static function getRawToken()
+    {
+        // Try to find the token in the `Authorization` header.
+        $headers = getallheaders();
+        if (!empty($headers['Authorization'])) {
+            $authorizationHeader = $headers['Authorization'];
+            if (is_string($authorizationHeader) && strpos($authorizationHeader, Tokens::HEADER_KEY) !== false) {
+                // The format for including the token in the header is slightly different then when included as a get or
+                // post parameter. Here the value will be in the form: `Bearer: <token>`
```

Comment:
```suggestion
                // post parameter. Here the value will be in the form: `Bearer <token>`
```
I thought that there isn't a colon between the Bearer and the value - please confirm